### ISSUE TITLE
Commands involving User Namespaces fail

### ISSUE THREAD

**Issue by Rpedvin** [apptainer/singularity#2397]
**Singularity Version:** 3.0.1-71.g9b2e243 (Cloned on 11/20/18)

**Base OS:** Centos7

**Issue:** Any commands involving creating user namespaces (i.e. passing the `--userns` flag) fail to mount to loop device. This prevents the container image from being mounted, and thus the singularity process fails. This occurs with `singularity shell`, `singularity exec` and `singularity instance start`.

**Expected Behavior:** Singularity correctly creates a blank (or minimal) new user namespace and launches a container within it.

**Actual Behavior:**

	[root@localhost ~]# singularity shell --userns singularity_test.simg
	FATAL:   container creation failed: mount error: can't mount image /proc/self/fd/8: failed to find loop device: could not attach image file too loop device: operation not permitted

**Debug logs:**

	[root@localhost ~]# singularity -d shell --userns singularity_test.simg
	[....]
	DEBUG   [U=0,P=8150]       Init()                        Check for image format sif
	DEBUG   [U=0,P=8145]       init()                                    Receiving configuration from scontainer stage 1
	DEBUG   [U=0,P=8145]       init()                                    Wait completion of scontainer stage1
	VERBOSE [U=0,P=8145]       setup_userns()                            Create user namespace
	DEBUG   [U=65534,P=8145]   setup_userns()                            Write deny to set group file
	DEBUG   [U=65534,P=8145]   setup_userns()                            Write to GID map
	DEBUG   [U=65534,P=8145]   setup_userns()                            Write line '0 0 1' to gid_map
	DEBUG   [U=65534,P=8145]   setup_userns()                            Write to UID map
	VERBOSE [U=0,P=8145]       create_namespace()                        Create mount namespace
	[....]
	DEBUG   [U=0,P=8145]       addResolvConfMount()          Adding /etc/resolv.conf to mount list
	DEBUG   [U=0,P=8145]       addHostnameMount()            Skipping hostname mount, not virtualizing UTS namespace on user request
	DEBUG   [U=0,P=8145]       create()                      Mount all
	DEBUG   [U=0,P=8145]       mountGeneric()                Mounting tmpfs to /usr/local/var/singularity/mnt/session
	DEBUG   [U=0,P=8145]       CleanupContainer()            Cleanup container
	FATAL   [U=0,P=8145]       SMaster()                     container creation failed: mount error: can't mount image /proc/self/fd/8: failed to find loop device: could not attach image file too loop device: operation not permitted

This also occurs when attempting to pass the `--userns` flag to an instance:

	[root@localhost ~]# singularity instance start --userns singularity_test.simg test
	ERROR:   container cleanup failed: no instance found with name test
	FATAL:   container creation failed: mount error: can't mount image /proc/self/fd/9: failed to find loop device: could not attach image file too loop device: operation not permitted

	FATAL:   failed to start instance: exit status 1

	[root@localhost ~]# singularity -d instance start --userns singularity_test.simg test
	[....]
	DEBUG   [U=0,P=8219]       Init()                        Check for image format sif
	DEBUG   [U=0,P=8218]       init()                                    Receiving configuration from scontainer stage 1
	DEBUG   [U=0,P=8218]       init()                                    Wait completion of scontainer stage1
	VERBOSE [U=0,P=8218]       init()                                    Run as instance
	VERBOSE [U=0,P=8223]       setup_userns()                            Create user namespace
	DEBUG   [U=65534,P=8223]   setup_userns()                            Write deny to set group file
	DEBUG   [U=65534,P=8223]   setup_userns()                            Write to GID map
	DEBUG   [U=65534,P=8223]   setup_userns()                            Write line '0 0 1' to gid_map
	DEBUG   [U=65534,P=8223]   setup_userns()                            Write to UID map
	VERBOSE [U=0,P=8223]       create_namespace()                        Create mount namespace
	DEBUG   [U=0,P=8223]       init()                                    Create RPC socketpair for communication between scontainer and RPC server
	[....]
	DEBUG   [U=0,P=8223]       addLibsMount()                Checking for 'user bind control' in configuration file
	DEBUG   [U=0,P=8223]       addResolvConfMount()          Adding /etc/resolv.conf to mount list
	DEBUG   [U=0,P=8223]       addHostnameMount()            Skipping hostname mount, not virtualizing UTS namespace on user request
	DEBUG   [U=0,P=8223]       create()                      Mount all
	DEBUG   [U=0,P=8223]       mountGeneric()                Mounting tmpfs to /usr/local/var/singularity/mnt/session
	DEBUG   [U=0,P=8223]       CleanupContainer()            Cleanup container
	ERROR   [U=0,P=8223]       SMaster()                     container cleanup failed: no instance found with name test
	FATAL   [U=0,P=8223]       SMaster()                     container creation failed: mount error: can't mount image /proc/self/fd/9: failed to find loop device: could not attach image file too loop device: operation not permitted

	FATAL   [U=0,P=8213]       execStarter()                 failed to start instance: exit status 1

**Hypothesis:** This issue is caused by nonexistent mappings and permissions created by invoking a new user namespace.

**Evidence:**
When creating a new user namespace, particularly though `unshare`, you get the same behavior that singularity demonstrates through it's debug logs.

Example of `strace unshare -r ls`:

	[....]
	 unshare(CLONE_NEWUSER)                  = 0
	 open("/proc/self/setgroups", O_WRONLY)  = 3
	 write(3, "deny", 4)                     = 4
	 close(3)                                = 0
	 open("/proc/self/uid_map", O_WRONLY)    = 3
	 write(3, "0 0 1", 5)                    = 5
	 close(3)                                = 0
	 open("/proc/self/gid_map", O_WRONLY)    = 3
	 write(3, "0 0 1", 5)                    = 5
	 close(3)                                = 0
	 [....]

 `unshare -r ls` creates a new user namespace (as seen with the `unshare(CLONE_NEWUSER)` system call), and then maps the current user (in this case, root) and it's uid/gid mappings to the root user's uid/gid respectively. This creates a new user namespace, and then executes the last argument (in this case, `ls`) in that created namespace. If no command is specified, `unshare` creates a new shell instead.

Results of running `unshare -r ls -al` in the /home/ directory...
 
 	[testUser@localhost home]$ ls -al
 	total 0
 	drwxr-xr-x.  4 root     root      37 Oct 17 19:46 .
 	dr-xr-xr-x. 18 root     root     239 Oct 17 17:39 ..
 	drwx------.  2 testUser testUser  62 Oct 17 19:46 testUser
 	drwx------.  4 vagrant  vagrant   90 Oct 17 17:40 vagrant

	[testUser@localhost home]$ unshare -r ls -al
	 total 0
	 drwxr-xr-x.  4 nfsnobody nfsnobody  37 Oct 17 19:46 .
	 dr-xr-xr-x. 18 nfsnobody nfsnobody 239 Oct 17 17:39 ..
	 drwx------.  2 root      root       62 Oct 17 19:46 testUser
	 drwx------.  4 nfsnobody nfsnobody  90 Oct 17 17:40 vagrant

  Everything that is not owned by the user that invoked the `unshare -r` command now shows up as nfsnobody, and everything that the user does own now shows up as root.

When trying to access the nfsnobody files/directories, it fails with a "PERMISSION DENIED" error.

	[testUser@localhost home]$ unshare -r
	[root@localhost home]# ls -al
	total 0
	drwxr-xr-x.  4 nfsnobody nfsnobody  37 Oct 17 19:46 .
	dr-xr-xr-x. 18 nfsnobody nfsnobody 239 Oct 17 17:39 ..
	drwx------.  2 root      root       83 Oct 17 23:43 testUser
	drwx------.  4 nfsnobody nfsnobody  90 Oct 17 17:40 vagrant
	[root@localhost home]# cd vagrant/
	-bash: cd: vagrant/: Permission denied

This also occurs when mapping the root user to the root user.

	[root@localhost home]# ls -al
	total 0
	drwxr-xr-x.  4 root     root      37 Oct 17 19:46 .
	dr-xr-xr-x. 18 root     root     239 Oct 17 17:39 ..
	drwx------.  2 testUser testUser  83 Oct 17 23:43 testUser
	drwx------.  4 vagrant  vagrant   90 Oct 17 17:40 vagrant 
	[root@localhost home]# sudo echo "hello"
	hello
	[root@localhost home]# unshare -r
	[root@localhost home]# ls -al
	total 0
	drwxr-xr-x.  4 root      root       37 Oct 17 19:46 .
	dr-xr-xr-x. 18 root      root      239 Oct 17 17:39 ..
	drwx------.  2 nfsnobody nfsnobody  83 Oct 17 23:43 testUser
	drwx------.  4 nfsnobody nfsnobody  90 Oct 17 17:40 vagrant
	[root@localhost home]# cd testUser/
	-bash: cd: testUser/: Permission denied
	[root@localhost home]# sudo echo "hello"
	sudo: PERM_SUDOERS: setresuid(-1, 1, -1): Invalid argument
	sudo: no valid sudoers sources found, quitting
	sudo: unable to initialize policy plugin

... Now let's take a look at what Singularity is attempting to access when mounting an image.

	[root@localhost home]# ls -al /dev/loop*
	brw-rw----. 1 root disk  7,   0 Oct 17 18:05 /dev/loop0
	crw-rw----. 1 root disk 10, 237 Oct 17 17:38 /dev/loop-control

These are the loopback devices in a Centos7 vagrant instance, which are correctly owned by the root user and disk user group. When looking at this inside a new user namespace, this is the result:

	[root@localhost home]# unshare -r ls -al /dev/loop*
	brw-rw----. 1 root nfsnobody  7,   0 Oct 17 18:05 /dev/loop0
	crw-rw----. 1 root nfsnobody 10, 237 Oct 17 17:38 /dev/loop-control

Alternatively as a non-root user, the result is this:

	[testUser@localhost ~]$ unshare -r ls -al /dev/loop*
	brw-rw----. 1 nfsnobody nfsnobody  7,   0 Oct 17 18:05 /dev/loop0
	crw-rw----. 1 nfsnobody nfsnobody 10, 237 Oct 17 17:38 /dev/loop-control`

... Since Singularity takes care of creating namespaces BEFORE the image gets mounted, it would make sense that the user is getting a "operation not permitted" error.

So how do we fix this?

<br>

Here are a couple of potential problems with the current user namespace implementation:
1) *It uses the `unshare()` syscall*

    If singularity's goal using the `--userns` flag is solely to map the user to root, then using `unshare` will work. However, in order to fix the current way that user namespaces are handled without major code overhaul, there is a chance that it will require multiple mappings.

	Unfortunately, `unshare` is incapable of this feat.
    
	I created some test code in order to figure out exactly why this doesn't work. You can find my code here: <https://pastebin.com/wxwvsCS4> 

	When solely mapping one line, consisting of the current user to root, the mapping successfully executes. 

		[root@localhost home]# ls -alh
		total 0
		drwxr-xr-x.  5 root     root      50 Nov 15 21:53 .
		dr-xr-xr-x. 18 root     root     239 Oct 17 17:39 ..
		drwx------.  2 test2    test2     62 Nov 15 21:53 test2
		drwx------.  2 testUser testUser  83 Oct 17 23:43 testUser
		drwx------.  4 vagrant  vagrant  246 Nov 20 23:28 vagrant

		[root@localhost home]# /root/Namespace_changer/clone_vs_unshare/unshare_single_mapping
		PID before creating new namespace: 18521
		PID after creating new namespace: 18521
		Executing command "ls -alh"...
		total 0
		drwxr-xr-x.  5 root      nfsnobody  50 Nov 15 21:53 .
		dr-xr-xr-x. 18 root      nfsnobody 239 Oct 17 17:39 ..
		drwx------.  2 nfsnobody nfsnobody  62 Nov 15 21:53 test2
		drwx------.  2 nfsnobody nfsnobody  83 Oct 17 23:43 testUser
		drwx------.  4 nfsnobody nfsnobody 246 Nov 20 23:28 vagrant

	However, mapping more than one line fails.

		[root@localhost home]# /root/Namespace_changer/clone_vs_unshare/unshare_multi_mapping
		PID before creating new namespace: 18518
		PID after creating new namespace: 18518
		Failed to write to uid_map file.

	The difference in strings that were mapped consisted of one entry, with the single mapping string being `0 0 1\n` and the multi string being `0 0 1\n1002 1002 1\n`, where id 1002 is the test2 user. Both strings are valid in terms of mapping. So why does it fail?

	Looking to the user_namespaces manpage, <http://man7.org/linux/man-pages/man7/user_namespaces.7.html>, it clearly states...
	 >    "In order for a process to write to the /proc/[pid]/uid_map
	 >      (/proc/[pid]/gid_map) file, all of the following requirements must be
	 >      met:...
	 >      
	 > ... 4. One of the following two cases applies:
	 >
	 >
	 >    *  Either the writing process has the CAP_SETUID (CAP_SETGID)
	 >            capability in the parent user namespace.
	 >       +  No further restrictions apply: the process can make mappings
	 >               to arbitrary user IDs (group IDs) in the parent user names‐
	 >               pace.
	 >
  	 >      *  Or otherwise all of the following restrictions apply:
	 >
  	 >            +  The data written to uid_map (gid_map) must consist of a sin‐
	 >               gle line that maps the writing process's effective user ID
	 >               (group ID) in the parent user namespace to a user ID (group
	 >               ID) in the user namespace.
	 >
	 >            +  The writing process must have the same effective user ID as
	 >               the process that created the user namespace.
	 >
	 >            +  In the case of gid_map, use of the setgroups(2) system call
	 >               must first be denied by writing "deny" to the
	 >               /proc/[pid]/setgroups file (see below) before writing to
	 >               gid_map.
	 >
	 >    Writes that violate the above rules fail with the error EPERM."

	Evidently, since it is possible to map one line successfully, but not multiple lines, requirement 4 falls through to the last three limitations. 
    
    My current understanding of why using the `unshare` call breaks this rule is because `unshare` "unshares" the specified namespace from the current process's parent namespace. Noting that the process who invokes `unshare` also modifies it's own user namespace, the "writing process" no longer exists in the parent user namespace, and thus has no capabilities there. Because of this, it breaks the rule of requiring the `CAP_SETUID`/`CAP_SETGID` capability, and thus falls through. This imposes the limitation that only one line may be written to the new user namespace mapping files, and that the UID/GID associated with writing process is the only UID/GID that may be mapped. Hence, Singularity's current implementation of user namespaces will break when attempting to map more than one line. This can be fixed by replacing ‘unshare’ with ‘clone’, which will be reviewed later in this writeup.

2) *The namespaces are mapped prior to the container being mounted*
	
    When looking at the full debug logs for `singularity -d shell --userns singularity_test.simg` command, the namespaces of the container are initialized prior to the container being mounted, as well as several other operations that the program executes in preparation for launching the container. Evidently, with the current way Singularity uses `unshare` for the user namespaces, this is going to cause problems. With `unshare`, it is only possible to map one line to the mapping files. Thus, with the current usage of `unshare` and the current location of where the namespaces are being created, this particular implementation is not going to exhibit the expected behavior.    
    

<br>

**Potential Solutions:**

All of these solutions require the use of the `clone` system call instead of the `unshare` call. 

Clone creates a new child process off of the process it is invoked from, and can also clone a new process with a new namespace. Calling the `clone` command with the `CLONE_NEWUSER` flag will create the child process within a new user namespace. The parent process of this child can then write to the child process's namespace, which fulfills the previously mentioned 4th requirement of user namespaces. Because of this, clone can create a child within a new user namespace and allow the original process to *map multiple lines within that child’s new namespace*. This opens up a lot of avenues for resolving the `--userns` issue.

Here's a small program that I wrote to help demonstrate how to use clone for this process: <https://pastebin.com/Zi25hSXN>

	[root@localhost home]# /root/Namespace_changer/clone_vs_unshare/clone_single_mapping 
	Executing "ls -alh" in new user namespace.
	total 0
	drwxr-xr-x.  5 root      nfsnobody  50 Nov 15 21:53 .
	dr-xr-xr-x. 18 root      nfsnobody 239 Oct 17 17:39 ..
	drwx------.  2 nfsnobody nfsnobody  62 Nov 15 21:53 test2
	drwx------.  2 nfsnobody nfsnobody  83 Oct 17 23:43 testUser
	drwx------.  4 nfsnobody nfsnobody 220 Nov 21 19:52 vagrant
	[root@localhost home]# /root/Namespace_changer/clone_vs_unshare/clone_multi_mapping 
	Executing "ls -alh" in new user namespace.
	total 0
	drwxr-xr-x.  5 root      nfsnobody  50 Nov 15 21:53 .
	dr-xr-xr-x. 18 root      nfsnobody 239 Oct 17 17:39 ..
	drwx------.  2 test2     nfsnobody  62 Nov 15 21:53 test2
	drwx------.  2 nfsnobody nfsnobody  83 Oct 17 23:43 testUser
	drwx------.  4 nfsnobody nfsnobody 220 Nov 21 19:52 vagrant

However, do take note that you can only map multiple lines as root. When attempting to do so with another user, this happens:

	[vagrant@localhost home]$ ./vagrant/clone_multi_mapping 
	Failed to write to uid_map file.
	[vagrant@localhost home]$ Executing "ls -alh" in new user namespace.
	total 0
	drwxr-xr-x.  5 nfsnobody nfsnobody  50 Nov 15 21:53 .
	dr-xr-xr-x. 18 nfsnobody nfsnobody 239 Oct 17 17:39 ..
	drwx------.  2 nfsnobody nfsnobody  62 Nov 15 21:53 test2
	drwx------.  2 nfsnobody nfsnobody  83 Oct 17 23:43 testUser
	drwx------.  4 nfsnobody nfsnobody 247 Nov 21 20:24 vagrant
	^C

So, in order to map multiple lines with `clone`, then it should be invoked alongside some of the code in singularity that requires higher privileges. 

<br>

With all that being said, here are a couple ideas to work around this:
    
1) Relocation of current namespace handling

	Relocating where certain namespaces are handled within the codebase may resolve this issue. If my current model of how Singularity works is correct, then once the container is mounted the sinit process is spawned. Then, all subsequent actions within the container are spawned from sinit (things like exec, shell, etc). If the process of creating and adopting a new user namespace was enacted alongside the creation of the sinit process, then it may surpass the current issue of not being able to mount to the loop devices. However, this also has the potential of causing other problems due to incomplete mappings, and may be rather work-intensive to implement.
    
2) Cloning a temporary child to host the namespace, then adopting later on

	The thought behind this is that as long as a namespace has a process that operates within it, it will continue to exist. Thus, instead of unsharing the user namespace, clone a new and minimal process with the `CLONE_NEWUSER` flag. Then modify the namespace within the cloned process, and have it wait for the container to finalize. Afterwards, as one of the last steps of creating sinit, the sinit process joins the user namespace of the placeholder process. After that, the placeholder process closes, and the container finalizes. In theory, this would prevent a major overhaul of code (such as that in suggestion 1), while still allowing the user to map multiple user/group mappings within the namespace. 

*Forcing mappings to include the system users and groups does not work.*

Once you implement `clone`, you can force system mappings to properly align so that they map to themselves. A UID mapping like `0 $user 1\n1 1 999\n` would map the current user to root, and then map all system users (assuming that users with IDs 0-999 are system users) to themselves. And, in the case of the loopback devices, if you map all of the system GIDs at a 1-to-1 in a similar fashion. The original thought was that this may take away the issue without requiring a massive code change. 

I modified my previously-mentioned ‘clone_test.c’ code to map both UID and GID values, and mapped the string `0 $user 1\n1 1 999\n` to both uid_map and gid_map to test this. This is the result (vagrant is user 1000):

	[root@localhost clone_vs_unshare]# gcc clone_test.c 
	[root@localhost clone_vs_unshare]# ./a.out 
	Executing "ls -alh /home" prior to new namespace
	total 4.0K
	drwxr-xr-x.  5 root     root       50 Nov 15 21:53 .
	dr-xr-xr-x. 18 root     root      239 Oct 17 17:39 ..
	drwx------.  2 test2    test2      62 Nov 15 21:53 test2
	drwx------.  2 testUser testUser   83 Oct 17 23:43 testUser
	drwx------.  5 vagrant  vagrant  4.0K Nov 27 21:19 vagrant
	Child pid is: 3809
	Executing "ls -alh /home" within new namespace
	total 4.0K
	drwxr-xr-x.  5 root      root        50 Nov 15 21:53 .
	dr-xr-xr-x. 18 root      root       239 Oct 17 17:39 ..
	drwx------.  2 nfsnobody nfsnobody   62 Nov 15 21:53 test2
	drwx------.  2 nfsnobody nfsnobody   83 Oct 17 23:43 testUser
	drwx------.  5 nfsnobody nfsnobody 4.0K Nov 27 21:19 vagrant
	Executing "ls -alh /dev/loop*"
	brw-rw----. 1 root disk  7,   0 Nov 27 22:27 /dev/loop0
	brw-rw----. 1 root disk  7,   1 Nov 27 21:18 /dev/loop1
	crw-rw----. 1 root disk 10, 237 Nov 27 21:06 /dev/loop-control
	Executing "singularity -d instance start /root/singularity_test.simg test" in new user namespace.
	VERBOSE [U=0,P=3816]       print()                       Set messagelevel to: 5
	DEBUG   [U=0,P=3816]       init()                        PIPE_EXEC_FD value: 10
	VERBOSE [U=0,P=3816]       init()                        Container runtime
	VERBOSE [U=0,P=3816]       is_suid()                     Check if we are running as setuid
	DEBUG   [U=0,P=3816]       init()                        Overlay seems not supported by kernel
	DEBUG   [U=0,P=3816]       init()                        Read json configuration from pipe
	DEBUG   [U=0,P=3816]       init()                        Set child signal mask
	DEBUG   [U=0,P=3816]       init()                        Create socketpair for smaster communication channel
	DEBUG   [U=0,P=3816]       init()                        Wait C and JSON runtime configuration from scontainer stage 1
	DEBUG   [U=0,P=3817]       set_parent_death_signal()     Set parent death signal to 9
	VERBOSE [U=0,P=3817]       init()                        Spawn scontainer stage 1
	VERBOSE [U=0,P=3817]       startup()                     Execute scontainer stage 1
	DEBUG   [U=0,P=3817]       Stage()                       Entering scontainer stage 1
	DEBUG   [U=0,P=3817]       prepareRootCaps()             Root full capabilities
	DEBUG   [U=0,P=3817]       Init()                        Entering image format intializer
	DEBUG   [U=0,P=3817]       Init()                        Check for image format sif
	DEBUG   [U=0,P=3816]       init()                        Receiving configuration from scontainer stage 1
	DEBUG   [U=0,P=3816]       init()                        Wait completion of scontainer stage1
	VERBOSE [U=0,P=3816]       init()                        Run as instance
	VERBOSE [U=0,P=3820]       priv_escalate()               Get root privileges
	VERBOSE [U=0,P=3820]       create_namespace()            Create mount namespace
	DEBUG   [U=0,P=3820]       init()                        Create RPC socketpair for communication between scontainer and RPC server
	VERBOSE [U=0,P=3820]       init()                        Create pid namespace
	VERBOSE [U=0,P=3820]       init()                        Spawn smaster process
	DEBUG   [U=0,P=1]          set_parent_death_signal()     Set parent death signal to 9
	VERBOSE [U=0,P=1]          init()                        Spawn scontainer stage 2
	VERBOSE [U=0,P=1]          create_namespace()            Create ipc namespace
	VERBOSE [U=0,P=1]          create_namespace()            Create mount namespace
	VERBOSE [U=0,P=2]          init()                        Spawn RPC server
	VERBOSE [U=0,P=3820]       startup()                     Execute smaster process
	VERBOSE [U=0,P=2]          startup()                     Serve RPC requests
	DEBUG   [U=0,P=3820]       setupSessionLayout()          Attempting to use overlayfs (enable overlay = try)
	DEBUG   [U=0,P=3820]       setupOverlayLayout()          Creating overlay SESSIONDIR layout
	DEBUG   [U=0,P=3820]       addRootfsMount()              Mount rootfs in read-only mode
	DEBUG   [U=0,P=3820]       addRootfsMount()              Mounting block [squashfs] image: /root/singularity_test.simg
	DEBUG   [U=0,P=3820]       addKernelMount()              Checking configuration file for 'mount proc'
	DEBUG   [U=0,P=3820]       addKernelMount()              Adding proc to mount list
	DEBUG   [U=0,P=3820]       addKernelMount()              Checking configuration file for 'mount sys'
	DEBUG   [U=0,P=3820]       addKernelMount()              Adding sysfs to mount list
	DEBUG   [U=0,P=3820]       addDevMount()                 Checking configuration file for 'mount dev'
	DEBUG   [U=0,P=3820]       addDevMount()                 Adding dev to mount list
	DEBUG   [U=0,P=3820]       addHostMount()                Not mounting host file systems per configuration
	VERBOSE [U=0,P=3820]       addBindsMount()               Found 'bind path' = /etc/localtime, /etc/localtime
	VERBOSE [U=0,P=3820]       addBindsMount()               Found 'bind path' = /etc/hosts, /etc/hosts
	DEBUG   [U=0,P=3820]       addHomeStagingDir()           Staging home directory (/root) at /usr/local/var/singularity/mnt/session/root
	DEBUG   [U=0,P=3820]       addHomeMount()                Adding home directory mount [/usr/local/var/singularity/mnt/session/root:/root] to list using layer: overlay
	DEBUG   [U=0,P=3820]       isLayerEnabled()              Using Layer system: overlay
	DEBUG   [U=0,P=3820]       addTmpMount()                 Checking for 'mount tmp' in configuration file
	DEBUG   [U=0,P=3820]       addScratchMount()             Not mounting scratch directory: Not requested
	DEBUG   [U=0,P=3820]       addLibsMount()                Checking for 'user bind control' in configuration file
	DEBUG   [U=0,P=3820]       addResolvConfMount()          Adding /etc/resolv.conf to mount list
	DEBUG   [U=0,P=3820]       addHostnameMount()            Skipping hostname mount, not virtualizing UTS namespace on user request
	DEBUG   [U=0,P=3820]       create()                      Mount all
	DEBUG   [U=0,P=3820]       mountGeneric()                Mounting tmpfs to /usr/local/var/singularity/mnt/session
	DEBUG   [U=0,P=3820]       CleanupContainer()            Cleanup container
	ERROR   [U=0,P=3820]       Master()                      container cleanup failed: no instance found with name test
	FATAL   [U=0,P=3820]       Master()                      container creation failed: mount error: can't mount image /proc/self/fd/9: failed to find loop device: could not attach image file too loop device: operation not permitted

	FATAL   [U=0,P=3812]       execStarter()                 failed to start instance: exit status 1

So forcing the mappings outright will not work, and another route will need to be taken.

<br>
    
**TL;DR:** The current way Singularity handles user namespaces isn't likely to work. Creating a new user namespace the way Singularity currently does will end up causing permission denied errors due to the inability to access files that appear as owned by unmapped users shown as "nfsnobody". As long as `unshare` is being used, Singularity will be unable to map multiple users. In my opinion, trying to switch to using `clone` for user-namespace handling would likely resolve the issue. By resolving this issue using `clone`, it would open the avenue for multiple lines of user mapping within the container. Changing how Singularity handles user namespaces may also assist in fixing the “fakeroot” issue, as specified in issue [#1924](https://github.com/sylabs/singularity/issues/1924).

I hope this helps with identifying the issue, and potentially resolving it in the future. If there’s anything else that I could do that would be useful, please let me know.

**Comment by cclerget** [comment]
Hi @Rpedvin and thank you for this detailed post. User namespace works only with directory (sandbox) image since loop devices requires root privileges, user namespace doesn't give those privileges even with a 1-to-1 mapping as root user, so it's not possible to use file image with user namespace.

If you use it like this :

```
singularity build --sandbox /tmp/mysandbox docker://alpine
singularity shell --userns /tmp/mysandbox
```

user namespace would work as expected.

**Comment by cclerget** [comment]
Hey @Rpedvin, the issue can be closed ?

**Comment by bauerm97** [comment]
@Rpedvin closing this issue as it should be resolved. If not, please re-open and ping me or @cclerget
